File: week4&5/Ariling/boj-2239-스도쿠/main.py
import sys
input = sys.stdin.readline
# pypy로 해야 통과... 왜인지는 모름..
arr = [list(map(int, input().rstrip())) for _ in range(9)]
def checkR(r, n) :
    for i in range(9):
        if arr[r][i] == n:
            return False
    return True
def checkC(c, n) :
    for i in range(9):
        if arr[i][c] == n:
            return False
    return True
def checkSquare(r, c, n):
    for i in range(3):
        for j in range(3):
            if arr[(r // 3) * 3 + i][(c // 3) * 3 + j] == n:
                return False
    return True
def find(n) :
    if n == len(blank):
        for i in arr:  # 출력 후
            print(''.join(map(str, i)))
        exit()  # 강제 종료
    r = blank[n][0]
    c = blank[n][1]
    for i in range(1, 10):
        if checkR(r, i) and checkC(c, i) and checkSquare(r, c, i):
            arr[r][c] = i
            find(n+1)
            arr[r][c] = 0
blank = []
for i in range(9):
    for j in range(9):
        if arr[i][j] == 0:
            blank.append((i, j))
find(0)

File: week4&5/Ariling/boj-2239-스도쿠/main.js
const fs = require("fs");
const [...input] = fs
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "./input.txt")
  .toString()
  .trim()
  .split("\n");

const arr = input.map((line) => line.split("").map(Number));

function solution(arr) {
  const blank = [];
  // 각 행 숫자 확인
  function row(x, n) {
    for (let i = 0; i < 9; i++) {
      if (n === arr[x][i]) {
        return false;
      }
    }
    return true;
  }
  // 각 열 숫자 확인
  function column(y, n) {
    for (let i = 0; i < 9; i++) {
      if (n === arr[i][y]) {
        return false;
      }
    }
    return true;
  }
  // 3 * 3 정사각형 숫자 확인
  function square(x, y, n) {
    for (let i = 0; i < 3; i++) {
      for (let j = 0; j < 3; j++) {
        if (n === arr[Math.floor(x / 3) * 3 + i][Math.floor(y / 3) * 3 + j]) {
          return false;
        }
      }
    }
    return true;
  }
  // 빈 칸 찾는 함수 -> 여러개이므로 일단은 돌려보자..
  function find(n) {
    if (n === blank.length) {
      const answer = arr.map((line) => line.join("")).join("\n");
      console.log(answer);
      //이래야 한다고 한다...
      process.exit();
    }
    const r = blank[n][0];
    const c = blank[n][1];
    for (let i = 1; i < 10; i++) {
      if (column(c, i) && row(r, i) && square(r, c, i)) {
        arr[r][c] = i;
        find(n + 1);
        arr[r][c] = 0;
      }
    }
  }
  for (let i = 0; i < arr.length; i++) {
    for (let j = 0; j < arr.length; j++) {
      if (arr[i][j] === 0) {
        blank.push([i, j]);
      }
    }
  }
  find(0);
}

solution(arr);
